[CINFRA-524] Add persisted state info and cleanup handler

maierlars opened the pull request:
### Scope & Purpose

Added persisted state info (only partially used) and a clean up handler. Used in the document store to drop the collection.

apetenchea reviewed (approved):
The code looks good to me!
We have a race when trying to drop the collection, which most of the time I can reproduce locally. Couldn't tell exactly where that is coming from, but I added some logging and it turns out that, when the collection drop blocks, the maintenance keeps seeing the shards, even after the replicated states have been dropped. I don't see how they could be somehow recreated quickly, so maybe the cleanup doesn't always do the job?

maierlars commented:
Yes, I saw that and I already fixed it in a follow up branch. The problem was: if the replicated log resigned faster than the replicated state, the replicated state would drop the core without destroying it. I fixed that in #17066 .
